File: Java/opentraceability/src/main/java/opentraceability/models/events/EPCISBaseDocument.java
package opentraceability.models.events;

import opentraceability.interfaces.IEvent;
import opentraceability.interfaces.IVocabularyElement;
import opentraceability.models.common.StandardBusinessDocumentHeader;
import opentraceability.models.identifiers.EPC;
import opentraceability.queries.EPCISQueryParameters;

import java.net.URI;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Stream;

public class EPCISBaseDocument {
    public EPCISVersion epcisVersion = null;
    public OffsetDateTime creationDate = null;
    public StandardBusinessDocumentHeader header = null;
    public ArrayList<IEvent> events = new ArrayList<IEvent>();
    public ArrayList<IVocabularyElement> masterData = new ArrayList<IVocabularyElement>();
    public HashMap<String, String> namespaces = new HashMap<String, String>();
    public ArrayList<String> contexts = new ArrayList<String>();
    public HashMap<String, String> attributes = new HashMap<String, String>();

    public <T extends IVocabularyElement> ArrayList<T> searchMasterData(Class<T> type) {
        ArrayList<T> result = new ArrayList<>();
        for (IVocabularyElement element : this.masterData) {
            if (type.isAssignableFrom(element.getClass())) {
                result.add((T) element);
            }
        }
        return result;
    }

    public <T extends IVocabularyElement> T searchMasterData(String id, Class<T> type) {
        for (IVocabularyElement element : this.masterData) {
            if (type.isAssignableFrom(element.getClass()) && element.getId().equals(id)) {
                return (T) element;
            }
        }
        return null;
    }

    public void merge(EPCISBaseDocument data) {
        for (IEvent e : data.events) {
            boolean found = false;
            for (IEvent e2 : this.events) {
                if (e.eventID.equals(e2.eventID)) {
                    if (e.errorDeclaration == null && e2.errorDeclaration != null) {
                        this.events.remove(e);
                        this.events.add(e2);
                    }
                    found = true;
                }
            }

            if (!found) {
                this.events.add(e);
            }

            for (IVocabularyElement element : data.masterData) {
                Stream<IVocabularyElement> singles = this.masterData.stream().filter(x -> x.getId().equals(element.getId()));
                if (singles.count() == 0) {
                    this.masterData.add(element);
                }
            }
        }
    }

    public List<IEvent> filterEvents(EPCISQueryParameters parameters) throws Exception {
        List<IEvent> result = new ArrayList<>();

        for (IEvent evt : this.events) {

            if (parameters.query.GE_eventTime != null) {
                if (evt.eventTime == null || evt.eventTime.isBefore(parameters.query.GE_eventTime)) {
                    continue;
                }
            }

            if (parameters.query.LE_eventTime != null) {
                if (evt.eventTime == null || evt.eventTime.isAfter(parameters.query.LE_eventTime)) {
                    continue;
                }
            }

            if (parameters.query.GE_recordTime != null) {
                if (evt.recordTime == null || evt.recordTime.isBefore(parameters.query.GE_recordTime)) {
                    continue;
                }
            }

            if (parameters.query.LE_recordTime != null) {
                if (evt.recordTime == null || evt.recordTime.isAfter(parameters.query.LE_recordTime)) {
                    continue;
                }
            }

            if (parameters.query.EQ_bizStep != null && !parameters.query.EQ_bizStep.isEmpty()) {

                if (!HasUriMatch(evt.businessStep, parameters.query.EQ_bizStep, "https://ref.gs1.org/cbv/BizStep-",
                        "urn:epcglobal:cbv:bizstep:")) {
                    continue;
                }
            }

            if (parameters.query.EQ_bizLocation != null && !parameters.query.EQ_bizLocation.isEmpty()) {
                boolean found = false;
                for (String eqBiz : parameters.query.EQ_bizLocation) {
                    if (evt.location != null && evt.location.gln != null
                            && eqBiz.toString().equalsIgnoreCase(evt.location.gln.toString())) {
                        found = true;
                        break;
                    }
                }
                if (!found) {
                    continue;
                }

            }

            if (parameters.query.MATCH_anyEPC != null && !parameters.query.MATCH_anyEPC.isEmpty()) {
                if (!HasMatch(evt, parameters.query.MATCH_anyEPC)) {
                    continue;
                }
            }

            if (parameters.query.MATCH_anyEPCClass != null && !parameters.query.MATCH_anyEPCClass.isEmpty()) {
                if (!HasMatch(evt, parameters.query.MATCH_anyEPCClass)) {
                    continue;
                }
            }

            if (parameters.query.MATCH_epc != null && !parameters.query.MATCH_epc.isEmpty()) {
                if (!HasMatch(evt, parameters.query.MATCH_epc, EventProductType.Reference, EventProductType.Child)) {
                    continue;
                }
            }

            if (parameters.query.MATCH_epcClass != null && !parameters.query.MATCH_epcClass.isEmpty()) {
                if (!HasMatch(evt, parameters.query.MATCH_epcClass, EventProductType.Reference,
                        EventProductType.Child)) {
                    continue;
                }
            }

            result.add(evt);
        }
        return result;
    }

    private boolean HasMatch(IEvent evt, List<String> epcs, EventProductType... allowedTypes) throws Exception {
        for (String epcMatchStr : epcs) {
            EPC epcMatch = new EPC(epcMatchStr);
            for (EventProduct product : evt.getProducts()) {
                if (allowedTypes.length == 0
                        || (product.Type != null && isInAllowedTypes(product.Type.toString(), allowedTypes))) {
                    if (epcMatch.Matches(product.EPC)) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    private boolean HasUriMatch(URI uri, List<String> filter, String prefix, String replacePrefix) {
        for (int i = 0; i < filter.size(); i++) {
            String bizStep = filter.get(i);
            URI u = URI.create(bizStep);
            if (u == null) {
                filter.set(i, replacePrefix + bizStep);
            } else if (bizStep.startsWith(prefix)) {
                filter.set(i, replacePrefix + bizStep.split("-")[2]);
            }
        }
        if (uri != null) {
            URI bizStep = URI.create(uri.toString());
            if (bizStep.toString().startsWith(prefix)) {
                bizStep = URI.create(
                        replacePrefix + uri.toString().substring(uri.toString().lastIndexOf("-") + 1));
            }
            List<URI> filterUris = new ArrayList<>();
            for (String s : filter) {
                filterUris.add(URI.create(s));
            }
            URI finalBizStep = bizStep;
            return filterUris.stream().anyMatch(u -> u.toString().equalsIgnoreCase(finalBizStep.toString()));
        } else {
            return false;
        }
    }

    private boolean isInAllowedTypes(String type, EventProductType... allowedTypes) {
        for (EventProductType _type : allowedTypes) {
            if (_type.toString().equals(type)) {
                return true;
            }
        }
        return false;
    }
}

File: Java/OpenTraceability/src/main/kotlin/opentraceability/models/events/epcisbasedocument.kt
package opentraceability.models.events

import opentraceability.interfaces.*
import opentraceability.models.common.StandardBusinessDocumentHeader
import opentraceability.models.identifiers.*
import opentraceability.queries.EPCISQueryParameters
import java.net.*
import java.time.OffsetDateTime

@Suppress("LocalVariableName", "PropertyName", "FunctionName")
open class EPCISBaseDocument {
    var epcisVersion: EPCISVersion? = null
    var creationDate: OffsetDateTime? = null
    var header: StandardBusinessDocumentHeader? = null
    var events: MutableList<IEvent> = mutableListOf()
    var masterData: MutableList<IVocabularyElement> =  mutableListOf()
    var namespaces: MutableMap<String, String> = mutableMapOf()
    var contexts: MutableList<String> = mutableListOf()
    var attributes: MutableMap<String, String> = mutableMapOf()

    inline fun <reified T : IVocabularyElement> searchMasterData(): MutableList<T> {
        return this.masterData.filterIsInstance<T>().toMutableList()
    }
    inline fun <reified T : IVocabularyElement> searchMasterData(id: String?): T? {
        return this.masterData.filterIsInstance<T>().firstOrNull { it.id == id }
    }

    open fun merge(data: EPCISBaseDocument) {
        data.events.forEach { e ->
            var found: Boolean = false;

            this.events.forEach { e2 ->
                if (e.eventID == e2.eventID) {
                    if (e.errorDeclaration == null && e2.errorDeclaration != null) {
                        this.events.remove(e);
                        this.events.add(e2);
                    }
                    found = true;
                }
            }

            if (!found) {
                this.events.add(e);
            }

            data.masterData.forEach { element ->
                var single = this.masterData.filter { x -> x.id == element.id }.single()

                if (single == null) {
                    this.masterData.add(element)
                }
            }
        }
    }

    fun filterEvents(parameters: EPCISQueryParameters): MutableList<IEvent> {
        val events = mutableListOf<IEvent>()

        for (evt in this.events) {
            // filter: GE_eventTime
            if (parameters.query.GE_eventTime != null) {
                if (evt.eventTime == null || evt.eventTime!! < parameters.query.GE_eventTime) {
                    continue
                }
            }

            // filter: LE_eventTime
            if (parameters.query.LE_eventTime != null) {
                if (evt.eventTime == null || evt.eventTime!! > parameters.query.LE_eventTime) {
                    continue
                }
            }

            // filter: GE_recordTime
            if (parameters.query.GE_recordTime != null) {
                if (evt.recordTime == null || evt.recordTime!! < parameters.query.GE_recordTime) {
                    continue
                }
            }

            // filter: LE_recordTime
            if (parameters.query.LE_recordTime != null) {
                if (evt.recordTime == null || evt.recordTime!! > parameters.query.LE_recordTime) {
                    continue
                }
            }

            // filter: EQ_bizStep
            if (parameters.query.EQ_bizStep != null && parameters.query.EQ_bizStep.isNotEmpty()) {
                if (!HasUriMatch(evt.businessStep, parameters.query.EQ_bizStep, "https://ref.gs1.org/cbv/BizStep-", "urn:epcglobal:cbv:bizstep:")) {
                    continue
                }
            }

            // filter: EQ_bizLocation
            if (parameters.query.EQ_bizLocation != null && parameters.query.EQ_bizLocation.isNotEmpty()) {
                if (evt.location?.gln == null || !parameters.query.EQ_bizLocation.map { it.toString().toLowerCase() }.contains(
                        evt.location!!.gln.toString().toLowerCase())) {
                    continue
                }
            }

            // filter: MATCH_anyEPC
            if (parameters.query.MATCH_anyEPC != null && parameters.query.MATCH_anyEPC.isNotEmpty()) {
                if (!HasMatch(evt, parameters.query.MATCH_anyEPC)) {
                    continue
                }
            }

            // filter: MATCH_anyEPCClass
            if (parameters.query.MATCH_anyEPCClass != null && parameters.query.MATCH_anyEPCClass.isNotEmpty()) {
                if (!HasMatch(evt, parameters.query.MATCH_anyEPCClass)) {
                    continue
                }
            }

            // filter: MATCH_epc
            if (parameters.query.MATCH_epc != null && parameters.query.MATCH_epc.isNotEmpty()) {
                if (!HasMatch(evt, parameters.query.MATCH_epc, EventProductType.Reference, EventProductType.Child)) {
                    continue
                }
            }

            // filter: MATCH_epcClass
            if (parameters.query.MATCH_epcClass != null && parameters.query.MATCH_epcClass.isNotEmpty()) {
                if (!HasMatch(evt, parameters.query.MATCH_epcClass, EventProductType.Reference, EventProductType.Child)) {
                    continue
                }
            }

            events.add(evt)
        }

        return events
    }

    fun HasMatch(evt: IEvent, epcs: MutableList<String>, vararg allowedTypes: EventProductType): Boolean {
        for (epcMatchStr in epcs) {
            val epcMatch = EPC(epcMatchStr)
            for (product in evt.products) {
                if (allowedTypes.isEmpty() || allowedTypes.contains(product.Type)) {
                    if (epcMatch.matches(product.EPC)) {
                        return true
                    }
                }
            }
        }
        return false
    }

    fun HasUriMatch(uri: URI?, filter: MutableList<String>, prefix: String, replacePrefix: String): Boolean {
        // make sure all of the EQ_bizStep are converted into URI format before comparing
        for (i in filter.indices) {
            val bizStep = filter[i]
            val u: URI? = URI.create(bizStep)
            if (u == null) {
                filter[i] = "$replacePrefix$bizStep"
            } else if (bizStep.startsWith(prefix)) {
                filter[i] = replacePrefix + bizStep.split("-").last()
            }
        }

        // we need to handle the various formats that the bizStep can occur in
        if (uri != null) {
            var bizStep = URI.create(uri.toString())
            if (bizStep.toString().startsWith(prefix)) {
                bizStep = URI.create("$replacePrefix${uri.toString().split("-").last()}")
            }

            val filterUris = filter.map { URI.create(it) }
            if (!filterUris.map { it.toString().toLowerCase() }.contains(bizStep.toString().toLowerCase())) {
                return false
            }
        } else {
            return false
        }

        return true
    }
}
